File: services/weatherService.js
const https = require('https');

function isWeatherQuery(message) {
    const msg = message.toLowerCase();
    const keywords = ['weather', 'temperature', 'forecast', 'climate', 'how hot', 'how cold', 'is it raining', 'rain in', 'temp in'];
    return keywords.some(kw => msg.includes(kw));
}

function fetchJson(url) {
    return new Promise((resolve, reject) => {
        const req = https.get(url, { headers: { 'User-Agent': 'Mozilla/5.0' }, timeout: 6000 }, (res) => {
            if (res.statusCode < 200 || res.statusCode >= 300) {
                return reject(new Error(`HTTP status ${res.statusCode}`));
            }
            let body = '';
            res.on('data', chunk => body += chunk);
            res.on('end', () => {
                try {
                    resolve(JSON.parse(body));
                } catch (e) {
                    reject(e);
                }
            });
        });
        req.on('error', reject);
        req.on('timeout', () => {
            req.destroy();
            reject(new Error('Request timed out'));
        });
    });
}

function extractLocation(message) {
    const msg = message.trim();
    const match = msg.match(/(?:weather|temperature|forecast|temp|climate)\s+(?:in|at|for|of)?\s*([a-zA-Z\s,]+)/i) ||
                  msg.match(/([a-zA-Z\s,]+)\s+(?:weather|temperature|forecast|climate)/i);

    if (match && match[1]) {
        const cleaned = match[1].replace(/\b(today|now|tomorrow|current|the|in|at|for|please|tell me|what is|whats|how is|is it)\b/gi, '').trim();
        if (cleaned.length > 1) return cleaned;
    }
    return 'Bhaktapur';
}

const WEATHER_EMOJIS = {
    'clear': '☀️',
    'clouds': '☁️',
    'rain': '🌧️',
    'drizzle': '🌦️',
    'thunderstorm': '🌩️',
    'snow': '❄️',
    'mist': '🌫️',
    'fog': '🌫️',
    'haze': '🌫️'
};

async function getWeatherFromOpenWeatherMap(city, apiKey) {
    const url = `https://api.openweathermap.org/data/2.5/weather?q=${encodeURIComponent(city)}&appid=${apiKey}&units=metric`;
    const data = await fetchJson(url);

    if (!data || data.cod != 200) return null;

    const placeName = `${data.name}, ${data.sys?.country || ''}`;
    const mainCondition = data.weather?.[0]?.main || 'Clear';
    const description = data.weather?.[0]?.description || mainCondition;
    const emoji = WEATHER_EMOJIS[mainCondition.toLowerCase()] || '🌤️';

    const tempC = data.main.temp.toFixed(1);
    const tempF = ((data.main.temp * 9/5) + 32).toFixed(1);
    const feelsC = data.main.feels_like.toFixed(1);
    const humidity = data.main.humidity;
    const windKmH = (data.wind.speed * 3.6).toFixed(1);

    return `### ${emoji} Weather for **${placeName}**\n` +
           `- **Condition:** ${description.charAt(0).toUpperCase() + description.slice(1)} ${emoji}\n` +
           `- **Temperature:** ${tempC}°C (${tempF}°F) | Feels like: ${feelsC}°C\n` +
           `- **Humidity:** ${humidity}%\n` +
           `- **Wind Speed:** ${windKmH} km/h\n` +
           `*Data provided in real-time by OpenWeatherMap.*`;
}

async function getWeatherFromOpenMeteo(city) {
    const geoUrl = `https://geocoding-api.open-meteo.com/v1/search?name=${encodeURIComponent(city)}&count=1&language=en&format=json`;
    const geoData = await fetchJson(geoUrl);

    if (!geoData.results || geoData.results.length === 0) return null;

    const location = geoData.results[0];
    const lat = location.latitude;
    const lon = location.longitude;
    const placeName = `${location.name}${location.country ? ', ' + location.country : ''}`;

    const weatherUrl = `https://api.open-meteo.com/v1/forecast?latitude=${lat}&longitude=${lon}&current_weather=true`;
    const weatherData = await fetchJson(weatherUrl);

    if (!weatherData.current_weather) return null;

    const curr = weatherData.current_weather;
    const tempC = curr.temperature;
    const tempF = ((tempC * 9/5) + 32).toFixed(1);

    return `### 🌤️ Weather for **${placeName}**\n` +
           `- **Temperature:** ${tempC}°C (${tempF}°F)\n` +
           `- **Wind Speed:** ${curr.windspeed} km/h\n` +
           `*Data provided in real-time by Open-Meteo.*`;
}

async function getWeatherData(message) {
    const city = extractLocation(message);
    const apiKey = process.env.WEATHER_API_KEY;

    // 1. Try OpenWeatherMap API
    if (apiKey) {
        try {
            const owmResult = await getWeatherFromOpenWeatherMap(city, apiKey);
            if (owmResult) return owmResult;
        } catch (e) {
            console.error('OpenWeatherMap API error:', e.message);
        }
    }

    // 2. Fallback to Open-Meteo API
    try {
        const omResult = await getWeatherFromOpenMeteo(city);
        if (omResult) return omResult;
    } catch (e) {
        console.error('Open-Meteo API error:', e.message);
    }

    return `Live weather information for "${city}" is currently unavailable right now. Please try again later.`;
}

module.exports = {
    isWeatherQuery,
    getWeatherData
};

File: backend/services/weather_service.py
"""
services/weather_service.py
Fetches live weather using OpenWeatherMap API with Open-Meteo fallback.
Uses api_manager for HTTP requests.
"""
from __future__ import annotations

import logging
import os
from typing import Optional

from ..utils.api_manager import get
from ..config import settings

logger = logging.getLogger(__name__)

_WEATHER_EMOJIS = {
    "clear": "☀️",
    "clouds": "☁️",
    "rain": "🌧️",
    "drizzle": "🌦️",
    "thunderstorm": "🌩️",
    "snow": "❄️",
    "mist": "🌫️",
    "fog": "🌫️",
    "haze": "🌫️",
    "smoke": "🌫️",
    "dust": "💨",
    "tornado": "🌪️",
}


def _extract_city(message: str) -> str:
    import re
    patterns = [
        r"weather\s+(?:in|at|for|of)?\s*([a-zA-Z\s,]+?)(?:\?|$|\s*please)",
        r"temperature\s+(?:in|at|for)?\s*([a-zA-Z\s,]+?)(?:\?|$)",
        r"forecast\s+(?:in|at|for)?\s*([a-zA-Z\s,]+?)(?:\?|$)",
        r"([a-zA-Z\s,]+?)\s+weather",
    ]
    for pattern in patterns:
        match = re.search(pattern, message, re.IGNORECASE)
        if match:
            city = match.group(1).strip()
            noise = ["today", "now", "current", "the", "please", "tell me", "what is", "how is", "is it"]
            for word in noise:
                city = city.replace(word, "").strip()
            if len(city) > 1:
                return city
    return "Bhaktapur"


def _get_from_openweathermap(city: str, api_key: str) -> Optional[str]:
    try:
        resp = get(
            "https://api.openweathermap.org/data/2.5/weather",
            params={"q": city, "appid": api_key, "units": "metric"},
            timeout=settings.weather_timeout,
            retries=2,
        )
        data = resp.json()
        if data.get("cod") != 200:
            return None

        name = data["name"]
        country = data.get("sys", {}).get("country", "")
        cond = data["weather"][0]["main"].lower()
        desc = data["weather"][0]["description"].capitalize()
        emoji = _WEATHER_EMOJIS.get(cond, "🌤️")
        temp_c = data["main"]["temp"]
        feels_c = data["main"]["feels_like"]
        humidity = data["main"]["humidity"]
        wind = data["wind"]["speed"]
        wind_kmh = round(wind * 3.6, 1)

        return (
            f"{emoji} **{name}**: {desc}\n"
            f"🌡️ {temp_c:.1f}°C | Feels {feels_c:.1f}°C\n"
            f"💧 {humidity}% | 🌬️ {wind_kmh} km/h"
        )
    except Exception as exc:
        logger.warning("OpenWeatherMap error: %s", exc)
        return None


def _get_from_open_meteo(city: str) -> Optional[str]:
    try:
        geo_resp = get(
            "https://geocoding-api.open-meteo.com/v1/search",
            params={"name": city, "count": 1, "language": "en", "format": "json"},
            timeout=settings.weather_timeout,
            retries=2,
        )
        geo = geo_resp.json()
        results = geo.get("results", [])
        if not results:
            return None

        loc = results[0]
        lat, lon = loc["latitude"], loc["longitude"]
        place = f"{loc['name']}{', ' + loc.get('country', '') if loc.get('country') else ''}"

        weather_resp = get(
            "https://api.open-meteo.com/v1/forecast",
            params={"latitude": lat, "longitude": lon, "current_weather": "true"},
            timeout=settings.weather_timeout,
            retries=2,
        )
        data = weather_resp.json()
        curr = data.get("current_weather")
        if not curr:
            return None

        temp_c = curr["temperature"]
        wind = curr["windspeed"]
        return f"🌤️ **{place}**: {temp_c}°C | Wind {wind} km/h"
    except Exception as exc:
        logger.warning("Open-Meteo error: %s", exc)
        return None


def get_weather(message: str) -> str:
    city = _extract_city(message)
    api_key = os.getenv("OPENWEATHER_API_KEY", settings.openweather_api_key)

    if api_key:
        result = _get_from_openweathermap(city, api_key)
        if result:
            return result

    result = _get_from_open_meteo(city)
    if result:
        return result

    return f"⚠️ Live weather for **{city}** is currently unavailable. Please try again later."


def is_weather_query(message: str) -> bool:
    keywords = ["weather", "temperature", "forecast", "climate", "how hot", "how cold", "is it raining", "temp in", "rain in"]
    return any(k in message.lower() for k in keywords)
